File: c++/parse.cpp
#include "error.cpp"
#include "objects.cpp"
// #include "lexer.cpp"
#include <algorithm>
#include <iostream>
#include <map>
#include <string>
#include <vector>

std::vector<std::string> allvar;

void vardec(std::vector<std::map<std::string, std::string>> &d,
            std::string type, Token &value) {
    std::map<std::string, std::string> d2;
    d2["type"] = "VariableDeclaration";
    d2["datatype"] = type;
    d2["value"] = value.val;
    d.push_back(d2);
}

void varassign(std::vector<std::map<std::string, std::string>> &d,
               std::string type, Token &name, Token &value, int line) {

    if (std::find(allvar.begin(), allvar.end(), name.val) ==
        allvar.end()) {
        try {
            throw TNLUndelcaredVariable(
                "\n\nVariable " + std::string(name.val) +
                " assigned value before declaration on line " +
                std::to_string(line) + ".");
        } catch (const std::exception &e) {
            std::cerr << e.what() << '\n';
        }
    }

    std::map<std::string, std::string> d2;
    d2["type"] = "VariableAssignment";
    d2["datatype"] = type;
    d2["name"] = name.val;
    d2["value"] = value.val;
    d.push_back(d2);
}

std::vector<std::map<std::string, std::string>>
parse(std::vector<Token> tokens) {
    static std::vector<std::map<std::string, std::string>> ast = {};

    std::vector<Token> n3t;
    std::vector<Token> p2t;

    bool skip2 = false;
    int line = 1;

    for (int k = 0; k < tokens.size(); k++) {

        try {
            n3t = {tokens[k], tokens[k + 1], tokens[k + 2], tokens[k + 3]};
        } catch (const std::exception &e) {
            try {
                n3t = {tokens[k], tokens[k + 1], tokens[k + 2]};
            } catch (const std::exception &e) {
                try {
                    n3t = {tokens[k], tokens[k + 1]};
                } catch (const std::exception &e) {
                    n3t = {tokens[k]};
                }
            }
        }

        p2t = {tokens[k - 2], tokens[k - 1]};

        if (tokens[k].type == "endstatement") {
            line++;
            continue;
        }
        if (skip2) {
            skip2 = false;
            continue;
        }

        if (n3t[0].type == "datatype" && n3t[1].type == "string") {
            allvar.push_back(n3t[1].val);
            if (n3t[0].val == "int") {
                vardec(ast, "int", n3t[1]);
            } else if (n3t[0].val == "str") {
                vardec(ast, "str", n3t[1]);
            }
            continue;
        }

        if (n3t[0].type == "string" && n3t[1].type == "assignment" &&
            n3t[2].type == "number") {
            varassign(ast, "int", n3t[0], n3t[2], line);
            continue;
        }

        if (n3t[0].type == "string" && n3t[1].type == "assignment" &&
            n3t[2].type == "punctuator") {
            skip2 = true;
            continue;
        }

        if (n3t[0].type == "punctuator" && n3t[1].type == "string" &&
            n3t[2].type == "punctuator" &&
            n3t[0].val == n3t[2].val) {
            Token rawstr("rawstring",
                         "\"" + std::string(n3t[1].val) + "\"");
            varassign(ast, "str", p2t[0], rawstr, line);
            continue;
        }

        if (n3t[0].type == "datatype" &&
            (std::find(allvar.begin(), allvar.end(), n3t[1].val) ==
             allvar.end()) &&
            n3t[1].type != "endstatement") {
            try {
                throw TNLSyntax(
                    "\n\nNon-alphabetic character in variable name on line " +
                    std::to_string(line) + ".");
            } catch (TNLSyntax &e) {
                std::cerr << e.what() << '\n';
            }
        }

        if (n3t[0].type == "datatype" &&
            (std::find(allvar.begin(), allvar.end(), n3t[2].val) ==
             allvar.end()) &&
            n3t[1].type == "number") {
            try {
                throw TNLSyntax(
                    "\n\nNon-alphabetic character in variable name on line " +
                    std::to_string(line) + ".");
            } catch (TNLSyntax &e) {
                std::cerr << e.what() << '\n';
            }
        }

        if (n3t[0].type == "datatype" && n3t[1].type != "string") {
            try {
                throw TNLSyntax(
                    "\n\nNon-alphabetic character in variable name on line " +
                    std::to_string(line) + ".");
            } catch (TNLSyntax &e) {
                std::cerr << e.what() << '\n';
            }
        }

        if (n3t[0].type == "punctuator" && n3t[1].type == "string" &&
            n3t[2].type != "punctuator") {
            try {
                throw TNLSyntax("\n\nUnclosed quotes on line " +
                                std::to_string(line) + ".");
            } catch (TNLSyntax &e) {
                std::cerr << e.what() << '\n';
            }
        }

        if (n3t[0].type == "punctuator" && n3t[1].type == "string" &&
            n3t[2].type == "punctuator" &&
            n3t[0].val != n3t[2].val) {
            try {
                throw TNLSyntax("\n\nUnmatched quotes on line " +
                                std::to_string(line) + ".");
            } catch (TNLSyntax &e) {
                std::cerr << e.what() << '\n';
            }
        }
    }

    return ast;
}

File: python/parse.py
from objects import Token
from lexer import token_types
import error


intvar: list[str] = []
strvar: list[str] = []
allvar: list[str] = []


def vardec(d: list, start: int, end: int, datatype: str, value: Token) -> None:
    d.append({
        'type': 'VariableDeclaration',
        'start': start,
        'end': end,
        'datatype': datatype,
        'value': value.val
    })


def assign(d: list, start: int, end: int, name: Token, value: Token, line: int) -> None:
    if name.val not in allvar:
        raise error.TNLUndeclaredVariable(f"\n\nVariable {name.val} assigned value before declaration on line {line}.")
    d2 = {
        'type': 'VariableAssignment',
        'start': start,
        'end': end,
        'name': name.val
    }
    
    try:
        d2['value'] = int(value.val)
    except ValueError:
        d2['value'] = value.val
    d.append(d2)


def parse(tokenlist: list[Token]):
    tokens = tokenlist
    ast = []

    tslice: int = 0
    # Next 3 tokens
    n3t: list[Token] = None
    # Previous 2 tokens
    p2t: list[Token] = None

    skip2: bool = False
    line = 1

    num_tokens = len(tokens)

    for k in range(num_tokens):
        global allvar, intvar, strvar
        tslice = k
        allvar = intvar+strvar

        n3t = tokens[tslice:min(tslice + 4, num_tokens)]

        p2t = [tokens[tslice-2], tokens[tslice-1]]

        if tokens[tslice].type == 'endstatement':
            line += 1
        if skip2:
            skip2 = False
            continue

        if n3t[0].type == 'datatype' and n3t[1].type == 'string':
            if n3t[0].val == 'int':
                intvar.append(n3t[1].val)
                vardec(ast, n3t[0].pos[0], n3t[1].pos[1], 'int', n3t[1])
            elif n3t[0].val == 'str':
                strvar.append(n3t[1].val)
                vardec(ast, n3t[0].pos[0], n3t[1].pos[1], 'str', n3t[1])
            continue

        if n3t[0].type == 'string' and n3t[1].type == 'assignment' and n3t[2].type == 'number':
            assign(ast, n3t[0].pos[0], n3t[2].pos[1], n3t[0], n3t[2], line)
            continue

        if n3t[0].type == 'string' and n3t[1].type == 'assignment' and n3t[2].type == 'punctuator':
            skip2 = True
            continue

        if n3t[0].type == 'punctuator' and n3t[1].type == 'string' and n3t[2].type == 'punctuator' and n3t[0].val == n3t[2].val:
            rawstr = Token('rawstring', '"' +
                           n3t[1].val+'"', (n3t[0].pos[0], n3t[2].pos[1]))
            assign(ast, rawstr.pos[0], rawstr.pos[1], p2t[0], rawstr, line)
            continue

        # Syntax checking

        if n3t[0].type == 'string' and (n3t[0].val in intvar or n3t[0].val in strvar) and n3t[1].type != 'endstatement':
            raise error.TNLSyntax(
                f"\n\nNon-alphabetic character in variable name on line {line}")
        
        if n3t[0].type == 'number' and (n3t[1].val in intvar or n3t[1].val in strvar) and n3t[1].type == 'string':
            raise error.TNLSyntax(
                f"\n\nNon-alphabetic character in variable name on line {line}")

        for i in token_types:
            if n3t[0].type == 'datatype' and n3t[1].type == i and i != 'string':
                raise error.TNLSyntax(
                    f"\n\nNon-alphabetic character in variable name on line {line}")

        if n3t[0].type == 'punctuator' and n3t[1].type == 'string' and not n3t[2].type == 'punctuator':
            raise error.TNLSyntax(f"\n\nUnclosed string on line {line}")

        if n3t[0].type == 'punctuator' and n3t[1].type == 'string' and n3t[2].type == 'punctuator' and n3t[0].val != n3t[2].val:
            raise error.TNLSyntax(f"\n\n Unmatched quotes on line {line}")

    return ast
